support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/MilestoneStepExecution.java

```diff
@@ -0,0 +1,443 @@
+
+import hudson.Extension;
+import hudson.XmlFile;
+import hudson.model.Executor;
+import hudson.model.InvisibleAction;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import jenkins.model.CauseOfInterruption;
+import jenkins.model.Jenkins;
+
+public class MilestoneStepExecution extends AbstractStepExecutionImpl {
+
+    private static final Logger LOGGER = Logger.getLogger(MilestoneStepExecution.class.getName());
+
+    @Inject(optional=true) private transient MilestoneStep step;
+    @StepContextParameter private transient Run<?,?> run;
+    @StepContextParameter private transient FlowNode node;
+    @StepContextParameter private transient TaskListener listener;
+
+    private static Map<String, Map<Integer, Milestone>> milestonesByOrdinalByJob;
+
+    @Override
+    public boolean start() throws Exception {
+        if (step.getLabel() != null) {
+            node.addAction(new LabelAction(step.getLabel()));
+        }
+        Integer ordinal = processOrdinal();
+        tryToPass(run, getContext(), ordinal);
+        return true;
+    }
+
+    private synchronized Integer processOrdinal() {
+        FlowGraphWalker walker = new FlowGraphWalker();
+        walker.addHead(node);
+        Integer previousOrdinal = null;
+        int parallelDetectionEnabled = 0;
+        for (FlowNode n : walker) {
+
+            if (parallelDetectionEnabled <= 0 && n.getAction(ThreadNameAction.class) != null) {
+                listener.getLogger().println("Milestone step found inside pararallel, it's not possible to grant ordering in this case.");
+                throw new MilestoneStepException("Using a milestone step inside parallel is not allowed");
+            }
+
+            if (n instanceof BlockEndNode) {
+                parallelDetectionEnabled++;
+            } else if (n instanceof BlockStartNode && !(n instanceof FlowStartNode)) {
+                parallelDetectionEnabled--;
+            }
+
+            OrdinalAction a = n.getAction(OrdinalAction.class);
+            if (a != null) {
+                previousOrdinal = a.ordinal;
+                break;
+            }
+        }
+        Integer nextOrdinal = 0;
+        if (previousOrdinal != null) {
+            nextOrdinal = previousOrdinal + 1;
+        }
+        node.addAction(new OrdinalAction(nextOrdinal));
+        return nextOrdinal;
+    }
+
+    private static class OrdinalAction extends InvisibleAction {
+        Integer ordinal;
+        public OrdinalAction(Integer ordinal) {
+            this.ordinal = ordinal;
+        }
+    }
+
+    // Used in tests only (where the static context is kept between tests somehow)
+    @Restricted(DoNotUse.class)
+    public static void clear() {
+        milestonesByOrdinalByJob = null;
+    }
+
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        throw new UnsupportedOperationException();
```

Comment:
Can just leave the body empty; ought not be called in any event, since we do not block.